perfaware/sim86/shared/Sim8086.cpp

```diff
@@ -303,62 +336,72 @@ int main(int ArgCount, char** Args)
     instruction_table Table;
     Sim86_Get8086InstructionTable(&Table);
     printf("8086 Instruction Instruction Encoding Count: %u\n", Table.EncodingCount);
-  
-    if (ArgCount > 1)
-    { 
-        for (int ArgIndex = 1; ArgIndex < ArgCount; ArgIndex++)
-        {
-            s16 Registers[RegisterCount] = {};
-            bool FlagArray[Flag_count] = {};
 
-            std::string OutputBuffer;
-            char* FileName = Args[ArgIndex];
-            std::ifstream File;
-            File.open(FileName, std::ifstream::binary | std::ifstream::in);
-            if (!File.good())
-            {
-                std::cout << "Error opening file " << FileName;
-                return -1;
-            }
-            
-            std::cout << "\n" << FileName << "\n";
-            std::vector<u8> Bytes((std::istreambuf_iterator<char>(File)), std::istreambuf_iterator<char>());
-            const u16 BytesRead = static_cast<u16>(Bytes.size());
-            u16& InstructionPointer = reinterpret_cast<u16&>(Registers[IPRegister]);
+    Memory = reinterpret_cast<u8*>(malloc(Megabyte));
 
-            while (InstructionPointer < BytesRead)
+    if (Memory)
+    {
+        if (ArgCount > 1)
+        {
+            for (int ArgIndex = 1; ArgIndex < ArgCount; ArgIndex++)
             {
-                instruction Decoded;
-                Sim86_Decode8086Instruction(BytesRead - InstructionPointer, &Bytes[0] + InstructionPointer, &Decoded);
+                s16 Registers[RegisterCount] = {};
+                bool FlagArray[Flag_count] = {};
 
-                if (Decoded.Op)
+                std::string OutputBuffer;
+                char* FileName = Args[ArgIndex];
+                std::ifstream File;
+                File.open(FileName, std::ifstream::binary | std::ifstream::in);
+                if (!File.good())
                 {
-                    InstructionPointer += static_cast<u16>(Decoded.Size);
-                    SimulateInstruction(Decoded, Registers, FlagArray);
+                    std::cout << "Error opening file " << FileName;
+                    return -1;
                 }
-                else
+
+                std::cout << "\n" << FileName << "\n";
+                std::vector<u8> Bytes((std::istreambuf_iterator<char>(File)), std::istreambuf_iterator<char>());
```

Comment:
Now that you have a memory, you should load your program into memory, not into some ad hoc vector that lives apart. That's how the computers actually work (remember also how Muratori showed the program in the first line of the image, he could do that because the program was in the same memory that the image was).

Reply:
Done!
c52bec4f7e58f9aa42f63d45afb60842ddc9863e

Reply:
1. `&Memory[0] + InstructionPointer` can be written more simply as `Memory + InstructionPointer`.
2. This part
```
            u8 Byte = static_cast<u8>(File.get());
            while (!File.fail())
            {
                Memory[InstructionPointer++] = Byte;
                Byte = static_cast<u8>(File.get());
            };
```
is a little bit awkward. It would be easier to gork with a for loop.
4. Instead of doing
```
            const u16 BytesRead = InstructionPointer;
            InstructionPointer = 0;
```
You can use BytesRead directly, you will spare yourself the need to set the IP to 0.

Reply:
THANK YOU, though I wonder why I didn't think about it in the first place, haha
a11353b1de42d41466e9182744ca18bd2fccce4b